File: YukioPlayer/Sources/YukioCore/CodexOpenChat.swift
import Foundation

/// Extracts selection/focus metadata from Codex's local desktop diagnostics.
/// Message bodies and window titles are not retained.
/// Only explicit view-activity events select a thread; background task events cannot select one.
/// This is an internal log format. Missing/changed data fails open: keep completion reminders.
public struct CodexOpenChatState {
    private var selected: [String: String] = [:]
    private var focusedWindow: String?
    private var focusObservedAt: Double = -.infinity

    private let timestampFormatter: ISO8601DateFormatter

    public init() {
        timestampFormatter = ISO8601DateFormatter()
        timestampFormatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
    }

    public mutating func ingest(_ line: String) {
        guard line.contains(" [electron-message-handler] "),
              let split = line.range(of: " [electron-message-handler] "),
              let timestamp = line.split(separator: " ").first,
              let date = timestampFormatter.date(from: String(timestamp)) else { return }
        let at = date.timeIntervalSince1970 * 1000
        let body = String(line[split.upperBound...])
        // Metadata is appended to diagnostic lines by the desktop logger.
        let fields = body.split(separator: " ").reduce(into: [String: String]()) { result, token in
            let pair = token.split(separator: "=", maxSplits: 1)
            if pair.count == 2 { result[String(pair[0])] = String(pair[1]) }
        }
        guard fields["rendererWindowAppearance"] == "primary",
              let window = fields["rendererWindowId"], Int(window) != nil,
              let focused = fields["rendererWindowFocused"], ["true", "false"].contains(focused),
              let visible = fields["rendererWindowVisible"], ["true", "false"].contains(visible) else { return }
        if at >= focusObservedAt {
            if focused == "true" && visible == "true" {
                focusedWindow = window
                focusObservedAt = at
            } else if focusedWindow == window {
                focusedWindow = nil
                focusObservedAt = at
            }
        }
        guard body.hasPrefix("thread_stream_view_activity_changed "),
              let id = fields["conversationId"], UUID(uuidString: id) != nil else { return }
        if fields["active"] == "true" {
            selected[window] = id
        } else if fields["active"] == "false", selected[window] == id {
            selected.removeValue(forKey: window)
        }
    }

    public func session(now: Double) -> String? {
        // A stale focus observation must not silently dismiss an unseen completion.
        guard now >= focusObservedAt, now - focusObservedAt <= 10_000,
              let window = focusedWindow else { return nil }
        return selected[window]
    }
}

/// Incremental reader. Restrict logs to the currently running Codex PID, so an old
/// process cannot suppress cards after restart. IO runs on the player's utility queue.
public final class CodexOpenChat {
    public let logsDirectory: URL
    private var processID: Int32?
    private var state = CodexOpenChatState()
    private var offsets: [URL: UInt64] = [:]
    private var partial: [URL: Data] = [:]

    public init(logsDirectory: URL = FileManager.default.homeDirectoryForCurrentUser
        .appendingPathComponent("Library/Logs/com.openai.codex")) {
        self.logsDirectory = logsDirectory
    }

    public func focusedSession(processID: Int32, now: Double) -> String? {
        if self.processID != processID {
            self.processID = processID
            state = CodexOpenChatState()
            offsets.removeAll()
            partial.removeAll()
        }
        let fm = FileManager.default
        guard let iterator = fm.enumerator(at: logsDirectory,
            includingPropertiesForKeys: [.fileSizeKey], options: [.skipsHiddenFiles]) else { return nil }
        let marker = "-\(processID)-t0-"
        let files = iterator.compactMap { $0 as? URL }.filter {
            $0.pathExtension == "log" && $0.lastPathComponent.hasPrefix("codex-desktop-")
                && $0.lastPathComponent.contains(marker)
        }.sorted { $0.path < $1.path }.suffix(8)
        guard !files.isEmpty else { return nil }
        for file in files {
            guard let values = try? file.resourceValues(forKeys: [.fileSizeKey]),
                  let size = values.fileSize, let handle = try? FileHandle(forReadingFrom: file) else { continue }
            defer { try? handle.close() }
            let end = UInt64(size)
            var start = offsets[file] ?? 0
            if start > end {
                // Truncation means the cached view can no longer be trusted.
                state = CodexOpenChatState()
                partial.removeValue(forKey: file)
                start = 0
            }
            guard start < end else { continue }
            // Bound startup/rotation work. If the view event is outside this tail,
            // no session is inferred; a subsequent navigation restores detection.
            let limit: UInt64 = 4 << 20
            let skipped = end - start > limit
            if skipped {
                start = end - limit
                partial.removeValue(forKey: file)
                state = CodexOpenChatState()
            }
            do {
                try handle.seek(toOffset: start)
                var data = partial[file] ?? Data()
                data.append(try handle.read(upToCount: Int(end - start)) ?? Data())
                offsets[file] = try handle.offset()
                if skipped, let newline = data.firstIndex(of: 10) { data.removeSubrange(...newline) }
                while let newline = data.firstIndex(of: 10) {
                    state.ingest(String(decoding: data[..<newline], as: UTF8.self))
                    data.removeSubrange(...newline)
                }
                partial[file] = data
            } catch { return nil }
        }
        return state.session(now: now)
    }
}

File: YukioWin/yukio/codex_open_chat.py
"""Selected Codex view from local diagnostics; unknown/stale data keeps reminders."""
from __future__ import annotations

import os
from pathlib import Path
from uuid import UUID
from .events import parse_timestamp


class CodexOpenChatState:
    def __init__(self):
        self.selected = {}
        self.selected_at = {}
        self.focused_window = None
        self.focus_at = float('-inf')

    def ingest(self, line):
        marker = ' [electron-message-handler] '
        if marker not in line:
            return
        at = parse_timestamp(line.split(' ', 1)[0])
        if at is None:
            return
        body = line.split(marker, 1)[1]
        fields = dict(token.split('=', 1) for token in body.split() if '=' in token)
        window = fields.get('rendererWindowId', '')
        focus = fields.get('rendererWindowFocused')
        visible = fields.get('rendererWindowVisible')
        if (fields.get('rendererWindowAppearance') != 'primary' or not window.isdigit()
                or focus not in ('true', 'false') or visible not in ('true', 'false')):
            return
        if at >= self.focus_at:
            if focus == visible == 'true':
                self.focused_window, self.focus_at = window, at
            elif self.focused_window == window:
                self.focused_window, self.focus_at = None, at
        if not body.startswith('thread_stream_view_activity_changed '):
            return
        session = fields.get('conversationId', '')
        try:
            UUID(session)
        except (ValueError, AttributeError):
            return
        if at < self.selected_at.get(window, float('-inf')):
            return
        if fields.get('active') == 'true':
            self.selected[window] = session
            self.selected_at[window] = at
        elif fields.get('active') == 'false':
            self.selected_at[window] = at
            if self.selected.get(window) == session:
                self.selected.pop(window, None)

    def session(self, now):
        if not 0 <= now - self.focus_at <= 10_000:
            return None
        return self.selected.get(self.focused_window)


class CodexOpenChat:
    def __init__(self, logs_directory=None):
        # Codex desktop's Windows file logger: LOCALAPPDATA/Codex/Logs/YYYY/MM/DD.
        self.logs_directory = Path(logs_directory) if logs_directory is not None else (
            Path(os.environ.get('LOCALAPPDATA') or Path.home() / 'AppData' / 'Local') / 'Codex' / 'Logs')
        self.process_id = None
        self.state = CodexOpenChatState()
        self.offsets = {}
        self.partial = {}

    def focused_session(self, process_id, now):
        if process_id != self.process_id:
            self.process_id = process_id
            self.state = CodexOpenChatState()
            self.offsets.clear()
            self.partial.clear()
        try:
            files = sorted(self.logs_directory.rglob(f'codex-desktop-*-{process_id}-t0-*.log'))[-8:]
            if not files:
                return None
            for path in files:
                end = path.stat().st_size
                start = self.offsets.get(path, 0)
                if start > end:
                    self.state = CodexOpenChatState()
                    self.partial.pop(path, None)
                    start = 0
                if start == end:
                    continue
                skipped = end - start > 4 << 20
                if skipped:
                    start = end - (4 << 20)
                    self.partial.pop(path, None)
                with path.open('rb') as stream:
                    stream.seek(start)
                    data = self.partial.get(path, b'') + stream.read(end - start)
                    self.offsets[path] = stream.tell()
                if skipped:
                    data = data.partition(b'\n')[2]
                lines = data.split(b'\n')
                self.partial[path] = lines.pop()
                for line in lines:
                    self.state.ingest(line.decode('utf-8', errors='replace'))
            return self.state.session(now)
        except OSError:
            return None
